s3/src/test/java/ch/cyberduck/core/sts/OidcAuthenticationTest.java

```diff
@@ -0,0 +1,154 @@
+        session.close();
+    }
+
+    @Test(expected = LoginFailureException.class)
+    public void testInvalidUserName() throws BackgroundException {
+        final Host host = new Host(profile, profile.getDefaultHostname(), new Credentials("WrongUsername", "rouser"));
+        final S3Session session = new S3Session(host);
+        session.open(Proxy.DIRECT, new DisabledHostKeyCallback(), new DisabledLoginCallback(), new DisabledCancelCallback());
+        session.login(Proxy.DIRECT, new DisabledLoginCallback(), new DisabledCancelCallback());
+        session.close();
+    }
+
+    @Test(expected = LoginFailureException.class)
+    public void testInvalidPassword() throws BackgroundException {
+        final Host host = new Host(profile, profile.getDefaultHostname(), new Credentials("rouser", "invalidPassword"));
+        final S3Session session = new S3Session(host);
+        session.open(Proxy.DIRECT, new DisabledHostKeyCallback(), new DisabledLoginCallback(), new DisabledCancelCallback());
+        session.login(Proxy.DIRECT, new DisabledLoginCallback(), new DisabledCancelCallback());
+        session.close();
+    }
+
+    @Test
+    public void testTokenRefresh() throws BackgroundException, InterruptedException {
+        final Host host = new Host(profile, profile.getDefaultHostname(), new Credentials("rawuser", "rawuser"));
+        final S3Session session = new S3Session(host);
+        session.open(Proxy.DIRECT, new DisabledHostKeyCallback(), new DisabledLoginCallback(), new DisabledCancelCallback());
+        session.login(Proxy.DIRECT, new DisabledLoginCallback(), new DisabledCancelCallback());
+
+        String firstAccessToken = host.getCredentials().getOauth().getIdToken();
+        String firstAccessKey = session.getClient().getProviderCredentials().getAccessKey();
+        String firstSessionToken = ((AWSSessionCredentials) session.getClient().getProviderCredentials()).getSessionToken();
+
+        Path container = new Path("cyberduckbucket", EnumSet.of(Path.Type.directory, Path.Type.volume));
+        assertTrue(new S3FindFeature(session, new S3AccessControlListFeature(session)).find(container));
+        Thread.sleep(1100 * 30);
+        assertTrue(new S3FindFeature(session, new S3AccessControlListFeature(session)).find(container));
+
+        assertNotEquals(firstAccessToken, host.getCredentials().getOauth().getIdToken());
+        assertNotEquals(firstAccessKey, session.getClient().getProviderCredentials().getAccessKey());
+        assertNotEquals(firstSessionToken, ((AWSSessionCredentials) session.getClient().getProviderCredentials()).getSessionToken());
+        session.close();
+    }
+
+    /** only use with the below specified changes in the keycloak config json file and run as separate test
+     * set config keycloak-realm.json:
+     *      "access.token.lifespan": "930"
+     *      "ssoSessionMaxLifespan": 1100,
+     */
+    @Test
+    @Ignore
```

Comment:
Moving discussion here from https://github.com/iterate-ch/cyberduck/pull/14893#discussion_r1289784236 

 @dkocher:
> Ideally we find a way to run these tests with different configurations in `keycloak-realm.json`.


@IsmCad:
> i My first thought was to use a second realm with these configurations, but that would lead to a second IDP config in MinIO which is not possible with the same address. As of now I do not know how to do this without maybe using additional containers, which would complicate the Testcontainers and the Docker config.


@chenkins 
> @dkocher I have drafted a way to run test with different configurations in `keycloak-realm.json`. `testSTSCredentialsExpiredValidOAuthToken` will take 15 minutes as the min STS duration is 900s see https://min.io/docs/minio/linux/developers/security-token-service/AssumeRoleWithWebIdentity.html. So unsure what it makes sense to finish it.
> 
> @IsmCadi I struggle to see the working of `testBucketRequestBeforeTokenExpiryFailsBecauseOfLatency` - can you re-phrase what you're testing and how to adjust the 28 secs? The test is red when I run locally.

Reply:
If an HTTP HEAD request to MinIO fails, the error code and error description are written to MinIO-specific error headers, since HTTP HEAD requests do not use a body. For example, a HEAD request is used to check whether an S3 bucket or folder is discoverable.
This issue came to light when a HEAD request uses expired STS credentials. Because of preemtively refresh tokens this case is only possible it the credentials are still valid but a few milliseconds before expire. Because of the latency in the network the request will be invalid when reaching the MinIO Service. But the sleep time needs to be ajusted according to the network latency.
Overall the test may be removed and the general question is how to handle the MinIO-specific HTTP-Headers when a HEAD-Request is failing.

Reply:
We are not interested in the special case of missing response body in HEAD requests where we have the S3 error code in the response header instead (handled in 56b96842) but rather if we can differentiate from the S3 API or STS API error if the OpenID Id token must be refreshed or not if missed in the preemptive handler.